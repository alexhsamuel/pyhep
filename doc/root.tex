\section{\module{hep.root} -- Root interface}

\declaremodule{extension}{hep.root}
\modulesynopsis{Root interface}

The \module{hep.root} module provides access from Python to features in
Root, a framework for high energy physics data management and analysis.
The applicable components of Root are included with \pyhep and need not
be provided separately.

\subsection{Root files}

Root includes an implementation of histograms and ntuples (which are
called ``trees'' in Root).  It implements a file format for storing
these in structured disk files.  \pyhep provides read and write access
to histograms and trees stored in Root files.

The \function{open} function access a Root file:

\begin{funcdesc}{open}{path\optional{, mode="r", purge_cycles=1}}
 Create or open a Root file at \var{path}.  

 \var{mode} is the access mode, analogous to Python's built-in
 \function{file} and \function{open} functions.  If \code{"r"}, opens an
 existing file for reading.  If \code{"r+"}, \code{"w"}, \code{"a"}, or
 \code{"a+"}, opens an existing file for reading and writing, or creates
 the file if it doesn't exist.  If \code{"w+"}, creates a file or
 overwrites an existing file.

 If \var{purge_cycles} is set to a true value and the file is open for
 writing, the entire file will be purged when it is closed.  
\end{funcdesc}

The return value from \function{open} is a \class{File} object.  The
file remains open until it is deleted.  

In \pyhep, all paths in Root files are specified relative to the root
directory of the Root file.  The root directory itself is represented by
an empty string, and the path separator is a forward slash.  Note that
there is no notion of working directory or relative paths.

Root files also support revision cycles; see the Root documentation for
details.  Unless \function{root.open} is invoked with
\code{purge_cycles=0} (or the file is opened read-only), \pyhep
``purges'' the entire file when the file is closed.  This removes cycles
except the most recent for each object in the file, whether or not the
object was written or revised in the current session.

\pyhep stores additional information along with tables and histograms in
Root files.  This information, called \emph{metadata}, consists of
useful information about histograms and tables that is not handled
directly by the Root file format.  \pyhep stores metadata in
subdirectories named ``_pyhep_metadata'' in the Root file.  These
subdirectories and their contents may be removed safely without injuring
the primary data in the histograms and tables they are associated with,
but doing so will delete configuration and annotations accessible from
\pyhep.

\subsubsection{\class{File} objects}

A \class{File} object has the following methods and attributes:

\begin{memberdesc}{writeable}
 True if the file is open for writing as well as reading.
\end{memberdesc}

\begin{methoddesc}{listdir}{\optional{path}}
 Return the contents of the directory specified by \var{path}.  The
 return value is a sequence of \code{(name, title, class_name)} tuples,
 where \var{name} is the object's name, \var{title} is the object's
 title, and \var{class_name} is the object's class name.
\end{methoddesc}

\begin{methoddesc}{mkdir}{path\optional{, title}}
 Create a new directory at \var{path}.  Optionally, a title may be
 specified in the \var{title} argument.
\end{methoddesc}

\begin{methoddesc}{rmdir}{path}
 Remove the directory at \var{path}.
\end{methoddesc}

\begin{methoddesc}{save}{path, object}
 Store a histogram \var{object} at \var{path}.  One- and two-dimensional
 histograms are supported.  
\end{methoddesc}

\begin{methoddesc}{load}{path}
 Load a histogram or connect to an ntuple at \var{path}.  If \var{path}
 is a one- or two-dimensional histogram, returns a histogram object with
 its contents.  If \var{path} is a tree, returns a table object (see
 \module{hep.table}) accessing it.
\end{methoddesc}

\begin{methoddesc}{createTable}{path, schema\optional{, title,
separate_branches=0, branch_name="default"}}
 Create a table as a tree at \var{path} in the Root file, using
 columns from \var{schema}, a \class{hep.table.Schema} instance.  

 By default, column values is stored in a leaves grouped in a single
 branch, named "default".  Another name may be specified for the branch
 with the \var{branch_name} argument.  Alternately, if
 \var{separate_branches} is true, each leaf is placed in a separate
 branch, whose name is the column name.

 These column types are supported: \constant{"int8"},
 \constant{"int16"}, \constant{"int32"}, \constant{"float32"}, and
 \constant{"float64"}.
\end{methoddesc}

\begin{methoddesc}{remove}{path}
 Remove the tree or histogram at \var{path}.  Use \method{rmdir} to
 remove a directory.
\end{methoddesc}

\begin{methoddesc}{normpath}{path}
 Return the canonical form of absolute path \var{path} in the Root
 file.
\end{methoddesc}

\begin{methoddesc}{join}{*components}
 Join together \var{components} to form an absolute path.
\end{methoddesc}

\begin{methoddesc}{split}{path}
 Return \code{(dir, name)}, where \var{dir} is the portion of \var{path}
 up to the last directory separator, and \var{name} is the remainder.
\end{methoddesc}

