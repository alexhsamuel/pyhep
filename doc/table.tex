\section{\module{hep.table} --- Tables}

\declaremodule{extension}{hep.table}
\modulesynopsis{Tables.}

A \pyhep \emph{table} is a flat database table with columns of fixed
type and storage size.  Tables are append-only: new rows may be added to
the end of the table, but rows may not be inserted elsewhere, nor can
rows be modified or deleted.  

Each table has a fixed \emph{schema}, which determines the data items
stored in the table and their types.  The schema includes an unordered
set of \emph{columns}, each identified by a name.  Each column has a
fixed data type.  A schema may also include \emph{expressions}, which
are derived quantities which are computed for each row from values of
columns and other expressions for that row.  

The table's data is an ordered sequence of \emph{rows}, each of which
contains exactly one value for each column in the table's schema.

Multiple table implementations are provided.  Implementations provide
different interfaces for creating and opening tables.  However, objects
representing tables and rows satisfy the same protocol across all
implementations.


\subsection{Schema objects}

An instance of \class{hep.table.Schema} represents a table schema.  

\begin{memberdesc}{columns}
 \readonly A sequence of columns in the schema.  Each element is a
 \class{Column} instance.  The order of the columns is unspecified.
\end{memberdesc}

\begin{memberdesc}{expressions}
 \readonly A sequence of expressions in the schema.  Each element is
 expression object (see \module{hep.expr}).  The order of the
 expressions is unspecified.
\end{memberdesc}

\begin{methoddesc}{addColumn}{name, type\optional{, **attributes}}
 Add a column to the sequence.  \var{name} is a character string, and
 \var{type} is a column type (see below).  If additional keyword
 arguments are provided, they are added to the column as attributes.
 New columns should not be added to a schema after the schema has been
 used to create a table.
\end{methoddesc}

\begin{methoddesc}{addExpression}{name, expression\optional{, **attributes}}
 Add an expression to the sequence.  \var{name} is a character string,
 and \var{expression} is an expression as a string or expression object.
 If additional keyword arguments are provided, they are added to the
 expression object as attributes.
\end{methoddesc}

An instance of \class{hep.table.Column} describes a column in a schema.
Do not instantiate this class directly; instead, use the
\method{addColumn} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The column's name.
\end{memberdesc}

\begin{memberdesc}{type}
 \readonly The column's data type.
\end{memberdesc}

The column's data type is specified by one of these names:
\begin{tableiii}{ccc}{code}{name}{C type}{Python Type}
  \lineiii{"int8"}{\code{signed char}}{\code{int}}
  \lineiii{"int16"}{\code{signed short}}{\code{int}}
  \lineiii{"int32"}{\code{signed int}}{\code{int}}
  \lineiii{"float32"}{\code{float}}{\code{float}}
  \lineiii{"float64"}{\code{double}}{\code{float}}
\end{tableiii}


An instance of \class{hep.table.Expression} describes a column in a
schema.  Do not instantiate this class directly; instead, use the
\method{addExpression} memthod of class \class{Schema}.

\begin{memberdesc}{name}
 \readonly The expressions's name.
\end{memberdesc}

\begin{memberdesc}{expression}
 \readonly An \class{Expression} instance; see \module{hep.expr}.
\end{memberdesc}

\begin{memberdesc}{formula}
 \readonly A string representation of the expression.
\end{memberdesc}



\subsection{Table objects}

A table object represents an in-memory table or a connection to an
externally-stored table.  In the former case, the table is erased when
the table object is deleted.  In the latter case, the connection is
closed but the table persists when the table object is deleted.

A table object satisfies the Python read-only sequence protocol.  The
sequence elements are the rows of the sequence.  Thus, the
\function{len} function returns the number of rows in the table, and
the \method{__getitem__} method extracts a row by its position in the
sequence. 

In addition, these methods are provided:

\begin{methoddesc}{append}{row\optional{, **kw_args}}
 (Read-write tables only.)  Appends a row to the end of the table.  The
 \var{row} must is a dictionary or other mapping object which maps
 column names to values.  Additional column values may be specified as
 keyword arguments.  If a column is specified both in \var{row} and in
 \var{**kw_args}, the value from latter is used.  A value must be
 specified for each column, and no other names may be specified.
 Returns the row index of the new row.
\end{methoddesc}

\begin{methoddesc}{__iter__}{}
 Returns an iterator over all rows in the table.
\end{methoddesc}

\begin{memberdesc}{rows}
 An interator over all rows in the table.
\end{memberdesc}

\begin{methoddesc}{select}{expr}
 Returns an iterator over rows in the table for which expression
 \var{expr} is true.  \var{expr} may be a string expression formula or
 an expression object.
\end{methoddesc}


\subsection{Row objects}

A row object may represent a row in a table, or a row that is compatible
with a table's schema (but not one of the rows in that table).

The row object satisfies the Python map protocol.  Each key is the name
of a column in the table's schema.  The corresponding map value is the
row's value for that column.  Row objects should be treated as
read-only; it is not possible to modify a table row by changing values
in a row object.

In addition, a row object provides these members and methods:

\begin{methoddesc}{asDict}{}
 Return a new Python dictionary of the column names and corresponding
 values in the row.
\end{methoddesc}

\begin{memberdesc}{index}
 \readonly The index of the row in the table.  The value of this
 attribute is undefined if the row was obtained from the table's
 \method{newRow} method and has not yet been passed to \method{append}.
\end{memberdesc}

\begin{memberdesc}{table}
 \readonly The table to which the row belongs.
\end{memberdesc}


\subsection{Standard implementation}

\pyhep's standard table implementation, in \module{hep.table}, uses a
simple, binary, on-disk representation of the table's data.  Table rows
are automatically loaded from disk as needed.

The \module{hep.table} module provides functions \function{create} and
\function{open} to create and open new tables, respectively.

\begin{funcdesc}{create}{filename, schema}
 Create a new table.  The table is stored in a file named by
 \var{filename}, which is created or overwritten.  The \var{schema}
 argument is a map from column names to \class{Column} instances,
 specifying the columns in the table.

 The return value is a table object, which is opened to the new, empty
 in write mode.
\end{funcdesc}

\begin{funcdesc}{open}{filename\optional{, mode="r"}}
 Open an existing table stored in the file named by \var{filename}.  The
 \var{mode} argument specifies the mode in which to open the table:
 \constant{"r"} to open the table in read-only mode, or \constant{"w"}
 to open the table in read-write mode.

 The return value is a table object.
\end{funcdesc}

