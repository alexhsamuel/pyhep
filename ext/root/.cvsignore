root.py
root_wrap.cc
root_wrap.doc
*Dict.cc
*Dict.h
*.pyc
depend.mk
ROOTSYS
