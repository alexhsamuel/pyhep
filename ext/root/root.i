//------------------------------------------------------------*- c++ -*-
//
// root.i
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

// SWIG interface file for Root libraries.

// -*- c++ -*-

%module root
%{
#include "TROOT.h"
#include "TKey.h"
#include "TDirectory.h"
#include "TFile.h"
#include "TLeaf.h"
#include "TLeafD.h"
#include "TLeafF.h"
#include "TLeafI.h"
#include "TList.h"
#include "TObject.h"
#include "TTree.h"
#include "TH1.h"
#include "TH2.h"
#include "TH3.h"
#include "TAxis.h"

class IndexError 
{
public:
  IndexError(int index) : index_(index) {};
  const int index_;
};

class StopIteration {};

%}

%inline %{
long 
getPointerValue(void* ptr) 
{
   return (long) ptr;
}
%}

%except(python) {
  try {
    $function
  }
  catch (IndexError e) {
    PyErr_Format(PyExc_IndexError, "%d", e.index_);
    return NULL;
  }
  catch (StopIteration) {
    PyErr_SetString(PyExc_StopIteration, "end of iterator");
    return NULL;
  }
}


class TObject
{
public:

  bool IsZombie() const;
  virtual const char* ClassName() const;
  virtual const char* GetName() const;
  virtual const char* GetTitle() const;
  virtual bool InheritsFrom(const char *classname) const;
  virtual void Print(const char* option) const;
  virtual int Write(const char* name=0, int option=0, int bufsize=0);

};


class TNamed
  : public TObject
{
public:

  virtual void SetName(const char* name);
  virtual void SetTitle(const char* title);

};


%pragma no_default

class TIterator
  : public TObject
{
public:

  virtual TObject* Next();

};

%addmethods TIterator {
  TIterator* __iter__() 
  {
    return self;
  };


  TObject* next()
  {
    TObject* object = self->Next();
    if (object == NULL)
      throw StopIteration();
    else
      return object;
  };
};



class TClass
  : public TObject
{
public:

  bool InheritsFrom(const char* cl) const;

};


class TKey
  : public TNamed
{
public:

  const char* GetClassName() const;
  short GetCycle() const;
  virtual TClass* IsA() const;

};


class TDirectory
  : public TNamed
{
public:

  virtual bool cd();
  virtual void Delete(const char* namecycle);
  virtual TObject* Get(const char* namecycle);
  TFile* GetFile() const;
  TKey* GetKey(const char* name, short cycle=9999);
  TList* GetList() const;
  TList* GetListOfKeys() const;
  TObject* GetMother() const;
  const char* GetPath() const;
  virtual TDirectory* mkdir(const char* name, const char* title);
  void Purge(short nKeep);
  
};


class TFile
  : public TDirectory
{
public:

  TFile(const char* fname, 
	char* option="", 
	const char* ftitle="", 
	int compress=1);

  virtual bool cd(const char *path = 0);
  virtual void Close(char* option="");
  virtual int Write(const char* name=0, int opt=0, int bufsiz=0);

};


class TList
  : public TObject
{
public:

  virtual int GetSize() const;
  virtual TObject* At(int idx) const;

};


%addmethods TList {
  int __len__() 
  {
    return self->GetSize();
  };


  TObject* __getitem__(int idx) 
  {
    if (idx < 0 || idx >= self->GetSize()) 
      throw IndexError(idx);
    else
      return self->At(idx);
  };
};



class TTree
  : public TObject
{
public:

  virtual int GetEntry(int entry=0, int getall=0);
  virtual double GetEntries() const;
  virtual TIterator* GetIteratorOnAllLeaves();
  virtual void Print();

};


class TLeaf
  : public TObject
{
public:
  
  virtual void SetAddress(void* add=0);

};

%addmethods TLeaf {
  void SetAddressInt(int add=0) 
  {
    self->SetAddress((void*) add);
  };
};


class TLeafD
  : public TLeaf
{
};


class TLeafF
  : public TLeaf
{
};


class TLeafI
  : public TLeaf
{
};


class TH1
  : public TNamed
{
public:

  virtual TAxis* GetXaxis();
  virtual TAxis* GetYaxis();
  virtual TAxis* GetZaxis();
  virtual double GetBinContent(int bin);
  virtual double GetBinError(int bin);
  virtual double GetEntries();
  virtual void SetBinContent(int bin, double content);
  virtual void SetBinError(int bin, double error);
  virtual void SetEntries(double entries);

};


class TH1F 
  : public TH1
{
public:

  TH1F(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1D
  : public TH1
{
public:

  TH1D(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1C
  : public TH1
{
public:

  TH1C(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH1S
  : public TH1
{
public:

  TH1S(const char* name, const char* title, int nbinsx, 
       double xlow, double xup);

};


class TH2
  : public TH1
{
public:

  virtual double GetBinContent(int binx, int biny);
  virtual double GetBinError(int binx, int biny);
  virtual void SetBinContent(int binx, int biny, double content);
  virtual void SetBinError(int binx, int biny, double error);

};


class TH2C
  : public TH2
{
public:

  TH2C(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2D
  : public TH2
{
public:

  TH2D(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2F
  : public TH2
{
public:

  TH2F(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH2S
  : public TH2
{
public:

  TH2S(const char* name, const char* title, int nbinsx, double xlow,
       double xup, int nbinsy, double ylow, double yup);

};


class TH3
  : public TH1
{
};


class TAxis
  : public TObject
{
public:

  int GetNbins();
  double GetXmin();
  double GetXmax();

};
