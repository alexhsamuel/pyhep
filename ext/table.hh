//----------------------------------------------------------------------
//
// table.hh
//
// Copyright 2003 by Alex Samuel.  All rights reserved.
//
//----------------------------------------------------------------------

/* Flat-file data table implementation and file format.  */

#ifndef __TABLE_HH__
#define __TABLE_HH__

//----------------------------------------------------------------------
// includes
//----------------------------------------------------------------------

#include <fcntl.h>
#include <string>
#include <sys/types.h>
#include <unistd.h>
#include <vector>

#include "value.hh"

//----------------------------------------------------------------------
// forward declarations
//----------------------------------------------------------------------

namespace table {

class Row;

//----------------------------------------------------------------------
// exception classes
//----------------------------------------------------------------------

class NoColumn 
{
public:

  NoColumn(const std::string& column_name);
  const std::string column_name_;

};


inline
NoColumn::NoColumn(const std::string& column_name)
  : column_name_(column_name)
{
}


class NotWritable
{
};


class NoRow
{
public:

  NoRow(int row_number) : row_number_(row_number) {}
  const int row_number_;

};


class FileError
{
public:

  FileError(const std::string& message);
  const std::string message_;

};


inline
FileError::FileError(const std::string& message)
  : message_(message)
{
}


class WrongColumnType
{
};


//----------------------------------------------------------------------
// types
//----------------------------------------------------------------------

enum ColumnType
{
  TYPE_NONE = -1,
  TYPE_BOOL = 0,
  TYPE_INT_8,
  TYPE_INT_16,
  TYPE_INT_32,
  TYPE_FLOAT_32,
  TYPE_FLOAT_64,
  TYPE_COMPLEX_64,
  TYPE_COMPLEX_128,
  TYPE_LAST
};


//----------------------------------------------------------------------

#if 0
template<typename TYPE>
struct ColumnTypeTemplate
{
  static ColumnType type;
};


template<>
ColumnType
ColumnTypeTemplate<bool>::type = TYPE_BOOL;


template<>
ColumnType
ColumnTypeTemplate<char>::type = TYPE_INT_8;


template<>
ColumnType
ColumnTypeTemplate<short>::type = TYPE_INT_16;


template<>
ColumnType
ColumnTypeTemplate<long>::type = TYPE_INT_32;


template<>
ColumnType
ColumnTypeTemplate<float>::type = TYPE_FLOAT_32;


template<>
ColumnType
ColumnTypeTemplate<double>::type = TYPE_FLOAT_64;
#else

template<typename TYPE>
ColumnType
getColumnType();

template<>
inline ColumnType
getColumnType<bool>()
{
  return TYPE_BOOL;
}


template<>
inline ColumnType
getColumnType<char>()
{
  return TYPE_INT_8;
}


template<>
inline ColumnType
getColumnType<short>()
{
  return TYPE_INT_16;
}


template<>
inline ColumnType
getColumnType<long>()
{
  return TYPE_INT_32;
}


template<>
inline ColumnType
getColumnType<float>()
{
  return TYPE_FLOAT_32;
}


template<>
inline ColumnType
getColumnType<double>()
{
  return TYPE_FLOAT_64;
}
#endif


//----------------------------------------------------------------------

class Column
{
public:

  Column(const std::string& name, const ColumnType type);
  Column(const Column& column);

  const std::string& getName() const;
  ColumnType getType() const;

private:

  std::string name_;
  ColumnType type_;

};


//----------------------------------------------------------------------

class Schema
{
public:

  Schema();
  virtual ~Schema();

  int getNumColumns() const;
  int whichColumn(const std::string& name) const throw (NoColumn);
  const Column& getColumn(int column_index) const;
  size_t getColumnOffset(int column_index) const;

  size_t getSize() const;

protected:

  struct ColumnRecord 
  {
    Column column_;
    size_t offset_;
  };

  std::vector<ColumnRecord> columns_;
  size_t size_;

};


class CustomSchema
  : public Schema
{
public:

  CustomSchema(size_t size=0);
  virtual ~CustomSchema();

  void addColumn(const Column& column, size_t offset);

};


class AutoSchema
  : public Schema
{
public:

  AutoSchema();
  virtual ~AutoSchema();

  void addColumn(const Column& column);

};


// FIXME: This is disabled; is it worth fixing?
#if 0
template<class STRUCT>
class StructSchema
  : public Schema
{
public:

  StructSchema();
  virtual ~StructSchema();

  template<typename FIELD_TYPE>
  void addColumn(const std::string& name, FIELD_TYPE STRUCT::* offset);

  void fillRow(Row* row, const STRUCT* structure) const;

private:

  void addColumn(const std::string& name, int offset, ColumnType type);

  std::vector<int> offsets_;

};
#endif


class Row
{
public:

  Row(const Schema* schema);
  ~Row();

  const Schema* getSchema() const
    { return schema_; }

  char* getBuffer() const
    { return data_; }

  Value getValue(int column_index) const throw (WrongColumnType);
  void setValue(int column_index, const Value& value);

private:

  const Schema* const schema_;
  char* data_;

};


class Table
{
protected:

  Table(const Schema* schema, bool writable)
    : schema_(schema), writable_(writable) {}

public:

  virtual ~Table();

  const Schema* getSchema() const
    { return schema_; }
  const bool isWritable() const
    { return writable_; }
  
  virtual int64_t getNumRows() const = 0;
  virtual void read(int64_t row_number, Row* row) = 0;
  virtual int append(const Row* row) = 0;
  virtual std::string getMetadata() const = 0;
  virtual void setMetadata(const std::string& data) = 0;

protected:

  const Schema* schema_;
  bool writable_;

};


class FileTable
  : public Table
{
public:

  ~FileTable();

  virtual int64_t getNumRows() const;
  virtual void read(int64_t row_number, Row* row);
  virtual int append(const Row* row);
  virtual std::string getMetadata() const;
  virtual void setMetadata(const std::string& data);

  static FileTable* create(const Schema* schema,
			   const std::string& path, 
			   mode_t mode=0666);
  static FileTable* open(const std::string& path, 
			 const std::string& mode);
  std::string getPath() const
    { return path_; }

protected:

  FileTable(const std::string& path, int flags=O_RDONLY);

private:

  const std::string path_;
  int fd_;

  off64_t first_row_offset_;
  off64_t row_size_;

  int64_t num_rows_;

};


//----------------------------------------------------------------------
// function declarations
//----------------------------------------------------------------------

extern size_t
getTypeSize(ColumnType type);

extern const char*
getTypeName(ColumnType type);

extern ColumnType
getTypeByName(const char* name);

//----------------------------------------------------------------------
// inline methods
//----------------------------------------------------------------------

inline
Column::Column(const std::string& name,
	       const ColumnType type)
  : name_(name),
    type_(type)
{
}


inline
Column::Column(const Column& column)
  : name_(column.name_),
    type_(column.type_)
{
}


inline const std::string&
Column::getName()
  const
{
  return name_;
}


inline ColumnType
Column::getType()
  const
{
  return type_;
}


//----------------------------------------------------------------------

inline
Schema::Schema()
  : size_(0)
{
}


inline int
Schema::getNumColumns()
  const
{
  return columns_.size();
}


inline const Column&
Schema::getColumn(int column_index)
  const
{
  return columns_[column_index].column_;
}


inline size_t
Schema::getColumnOffset(int column_index)
  const
{
  return columns_[column_index].offset_;
}


inline size_t
Schema::getSize()
  const
{
  return size_;
}


//----------------------------------------------------------------------

inline
CustomSchema::CustomSchema(size_t size)
{
  size_ = size;
}


inline
CustomSchema::~CustomSchema()
{
}


inline void
CustomSchema::addColumn(const Column& column,
			size_t offset)
{
  ColumnRecord record = { column, offset };
  columns_.push_back(record);
}


//----------------------------------------------------------------------

#if 0
template<class STRUCT>
inline
StructSchema<STRUCT>::StructSchema()
{
}


template<class STRUCT>
inline
StructSchema<STRUCT>::~StructSchema()
{
}


template<class STRUCT>
template<typename FIELD_TYPE>
inline void
StructSchema<STRUCT>::addColumn(const std::string& name,
				FIELD_TYPE STRUCT::* offset)
{
  addColumn(name, (int) &(((STRUCT*) NULL)->*offset),
	    getColumnType<FIELD_TYPE>());
}


template<class STRUCT>
inline void
StructSchema<STRUCT>::fillRow(Row* row,
			      const STRUCT* structure)
  const
{
  int num_columns = columns_.size();
  for (int c = 0; c < num_columns; ++c) {
    const char* field_address = ((const char*) structure) + offsets_[c];
    row->setValue(c, (void*) field_address);
  }
}


template<class STRUCT>
inline void
StructSchema<STRUCT>::addColumn(const std::string& name,
				int offset,
				ColumnType type)
{
  Column column(name, type);
  ColumnRecord record = { column, size_ };
  columns_.push_back(record);
  size_ += getTypeSize(column.getType());
  offsets_.push_back(offset);
}
#endif


//----------------------------------------------------------------------

inline int64_t
FileTable::getNumRows()
  const
{
  return num_rows_;
}


//----------------------------------------------------------------------

}  // namespace table


#endif  // #ifndef __TABLE_HH__
