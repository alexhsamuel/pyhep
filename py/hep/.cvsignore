*.pyc
*.pyo
config.py
*-install
pdt.pickle
