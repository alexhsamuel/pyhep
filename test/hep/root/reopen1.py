#-----------------------------------------------------------------------
# imports
#-----------------------------------------------------------------------

import hep.fs
import hep.root
import hep.hist
import hep.table 
from   hep.test import compare

#-----------------------------------------------------------------------
# clean up
#-----------------------------------------------------------------------

cwd = hep.fs.getdir(".", writable=True)
if "reopen1.root" in cwd:
    del cwd["reopen1.root"]

#-----------------------------------------------------------------------
# test
#-----------------------------------------------------------------------

# Construct a schema with three columns.
schema = hep.table.Schema(a="float32", b="float32")

# Create a new Root file.
root_file = hep.fs.getdir("reopen1.root", makedirs=True, writable=True)
# Create a tree in it.
table = hep.root.createTable("tree", root_file, schema, title="PyHEP test")
# Fill 100 random rows into the ntuple.
table.append(a=1, b=2)
# Release these to close the Root file.
del table, root_file

# Reopen the Root file.
root_file = hep.fs.getdir("reopen1.root", writable=True)
# Get the table.
table = root_file["tree"]
# Fill a row into the ntuple.
table.append(a=4, b=5)
# Release these to close the Root file.
del table, root_file

# Reopen the Root file.
root_file = hep.fs.getdir("reopen1.root")
# Get the table.
table = root_file["tree"]
# Make sure both rows are there.
compare(len(table), 2)
compare(table[0]["a"], 1)
compare(table[1]["b"], 5)
# Release these to close the Root file.
del table, root_file

