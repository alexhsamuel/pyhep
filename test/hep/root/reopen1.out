Traceback (most recent call last):
  File "reopen1.py", line 17, in ?
    del cwd["reopen1.root"]
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 111, in __delitem__
    subdir.__del(key)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 410, in __del
    directory = self.get(key, **args)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 162, in get
    return subdir._get(key, **kw_args)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 632, in _get
    return type[3](self, key, **kw_args)
  File "/home/samuel/dev/pyhep/py/hep/fs.py", line 772, in __get_root_file
    return hep.root.open(path, writable, purge_cycles, with_metadata)
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 667, in open
    return file.root
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 541, in <lambda>
    root = property(lambda self:
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 65, in __init__
    for tkey in [ _cast(tlist[i]) for i in range(len(tlist)) ]:
  File "/home/samuel/dev/pyhep/py/hep/root/__init__.py", line 600, in _cast
    parts = object.this.split("_")
AttributeError: 'PySwigObject' object has no attribute 'split'
